### ISSUE TITLE
how can I understand self.true_boxes 

### ISSUE THREAD

**Issue by leadcain84** [experiencor/keras-yolo2#286]
Thank you for source code.

I have built YOLOv2 scratch train env. on tensorflow only.

It is greatly helpful for me.

I try to understand to use self.true_boxes, but it is hard to me.

I can't understand why self.true_boxes is used, and which values are assigned.

It is used to get iou_scores between pred_areas and true_areas.

here is the code in the frontend.py
39 line :  self.true_boxes = Input(shape=(1, 1, 1, max_box_per_image, 4))
 . . .
69 line : output = Lambda(lambda args: args[0])([output, self.true_boxes])
71 line : self.model = Model([input_image, self.true_boxes], output)
. . .
147 line : true_xy = self.true_boxes[..., 0:2]
148 line : true_wh = self.true_boxes[..., 2:4]
...

I would appreciate anybody helping me.

**Comment by Axel13fr** [comment]
hi @leadcain84 , this tensor contains the ground truth boxes. They are input into the model because the loss is computed in the tensor flow graph directly to do it efficiently. 

To compute IOUs,  you need the position and dimensions of the boxes and the IOUs are used to associate the detections to ground truth boxes. From there, loss terms are computed.

For a better understanding, you may want to read the many articles about how YOLO loss is computed and then come back to this implementation (which does appear complex at first because it's all done in tensor flow...).

**Comment by italianconcerto** [comment]
> hi @leadcain84 , this tensor contains the ground truth boxes. They are input into the model because the loss is computed in the tensor flow graph directly to do it efficiently.
> 
> To compute IOUs, you need the position and dimensions of the boxes and the IOUs are used to associate the detections to ground truth boxes. From there, loss terms are computed.
> 
> For a better understanding, you may want to read the many articles about how YOLO loss is computed and then come back to this implementation (which does appear complex at first because it's all done in tensor flow...).

Excuse me, what's the difference between self.true_boxes and y_true then?
I don't get it
